File: code/reporter/src/app/utils/ip_utils.py
import re

import ipaddress


def cidr_to_ip_list(input_data: str) -> list | None:
    if not input_data or "/" not in input_data:
        return None

    try:
        range_ip_parse_result = ipaddress.ip_network(input_data, strict=False)

        if range_ip_parse_result:
            ip_list = [str(ip) for ip in range_ip_parse_result]
            return ip_list
    except Exception:
        return None
    return None


def get_ip_count_if_range(input_data: str) -> int:
    """
    * Get IP count if range

    Args:
        input_data (str): _description_

    Returns:
        int: _description_
    """
    if not input_data or "/" not in input_data:
        return 0

    try:
        ip_list = cidr_to_ip_list(input_data)
        if not ip_list:
            return 0

        return len(ip_list)
    except Exception:
        return 0
    return 0


def is_valid_cidr_by_regex(cidr_str: str) -> bool:
    # Define the regular expression pattern for CIDR notation
    pattern = (
        r"^(?:(?:25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|"
        r"[0-1]?[0-9][0-9]?)\/(?:[1-9]|[1-2][0-9]|3[0-2])$"
    )

    # Use the re.match function to check if the input matches the pattern
    return bool(re.match(pattern, cidr_str))

File: code/scanner-api/utils/iputils/iputils.go
package iputils

import (
	"errors"
	"math"
	"math/big"
	"net"
	"regexp"
	"strings"
)

func GetIPCountIfRange(cidr string) (int, error) {

	_, ipnet, err := net.ParseCIDR(cidr)
	if err != nil {
		return 0, err
	}

	// Get the number of bits in the network mask
	ones, _ := ipnet.Mask.Size()

	// Calculate the number of host bits
	hostBits := 32 - ones

	// Calculate the number of usable IP addresses (excluding network and broadcast addresses)
	totalIPs := int(math.Pow(2, float64(hostBits)))

	return totalIPs, nil
}

func IsValidCIDRByRegex(cidrStr string) bool {
	// Define the regular expression pattern for CIDR notation
	pattern := `^(?:(?:25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\/(?:[1-9]|[1-2][0-9]|3[0-2])$`

	// Use the regexp.MatchString function to check if the input matches the pattern
	match, _ := regexp.MatchString(pattern, cidrStr)
	return match
}

func ConvertIPRangeToIPSize(ipRange string) (*big.Int, error) {
	if ipRange == "" {
		return nil, nil
	}

	ipRange = strings.TrimSpace(ipRange)

	if strings.Contains(ipRange, "-") {
		return nil, errors.New("Invalid Target")
	}

	if strings.Contains(ipRange, "/") {

		totalIPs, err := countCIDR(ipRange)
		if err != nil {
			return nil, err
		}

		return totalIPs, nil
	}

	return nil, nil
}

// CIDRRangeSize calculates the total number of IP addresses in a CIDR range.
func countCIDR(cidr string) (*big.Int, error) {
	var totalIPs *big.Int
	_, ipNet, err := net.ParseCIDR(cidr)
	if err != nil {
		return totalIPs, err
	}

	// Get the number of bits in the mask
	ones, bits := ipNet.Mask.Size()

	// Calculate the number of possible IPs
	// The formula is 2^(bits - ones)
	totalIPs = big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(bits-ones)), nil)

	return totalIPs, nil
}
